app/login.go

```diff
@@ -55,6 +58,26 @@ func (a *App) AuthenticateUserForLogin(id, loginId, password, mfaToken string, l
 		return nil, err
 	}
 
+	// CWS login allow to use the one-time token to login the users when they're redirected to their
+	// installation for the first time
+	if IsCWSLogin(a, cwsToken) {
+		if err = checkUserNotBot(user); err != nil {
+			return nil, err
+		}
+		sessions, errSessions := a.GetSessions(user.Id)
+		if errSessions != nil {
+			return nil, errSessions
+		}
+		if len(sessions) > 0 {
+			return nil, model.NewAppError("AuthenticateUserForLogin",
+				"api.user.login_by_cws.invalid_token.app_error", nil, "", http.StatusBadRequest)
+		}
+		token, ok := os.LookupEnv(cwsTokenEnv)
+		if ok && token == cwsToken {
```

Comment:
I think we need to use [`subtle.ConstantTimeCompare`](https://golang.org/pkg/crypto/subtle/#ConstantTimeCompare) here to avoid a timing side-channel.